src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -660,49 +667,45 @@ private LocalSymbol GetDeclaredLocal(CSharpSyntaxNode declarationSyntax, SyntaxT
+                        return GetAdjustedLocalSymbol((SourceLocalSymbol)local);
                     }
                 }
             }
 
             return null;
         }
 
-        internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int position)
+#nullable enable
+        internal override LocalSymbol GetAdjustedLocalSymbol(SourceLocalSymbol local)
         {
-            Debug.Assert(local is SourceLocalSymbol);
-            LocalSymbol adjustedLocal;
-            if (Compilation.NullableSemanticAnalysisEnabled)
-            {
-                if (!_analyzedVariableTypesOpt.TryGetValue(local, out adjustedLocal))
-                {
-                    var types = GetSnapshotManager().GetVariableTypesForPosition(position);
-
-                    // If the local was not inferred, it does not get an entry in this dictionary. Save the local mapped
-                    // to itself to avoid needing to enter this code path in the future.
-                    if (types.TryGetValue(local, out TypeWithAnnotations type))
-                    {
-                        adjustedLocal = _analyzedVariableTypesOpt.GetOrAdd(local, ((SourceLocalSymbol)local).WithAnalyzedType(type));
-                    }
-                    else
-                    {
-                        _analyzedVariableTypesOpt.TryAdd(local, local);
-                        adjustedLocal = local;
-                    }
-                }
-            }
-            else
-            {
-                adjustedLocal = local;
-            }
+            return GetRemappedSymbol<LocalSymbol>(local);
+        }
 
-            return adjustedLocal;
+        private LocalFunctionSymbol GetDeclaredLocalFunction(LocalFunctionStatementSyntax declarationSyntax)
+        {
+            // If nullable semantic analysis is enabled, we need to potentially reinfer the containing symbol
+            // of a local function, as it could be in a reinferred lambda. For an example, see
+            // GetSymbolInfo_NestedLambdaReinference_LocalFunctionInLambda
+            var originalSymbol = GetDeclaredLocalFunction(this.GetEnclosingBinder(GetAdjustedNodePosition(declarationSyntax)), declarationSyntax.Identifier);
+            return GetRemappedSymbol(originalSymbol);
         }
 
-        private LocalFunctionSymbol GetDeclaredLocalFunction(LocalFunctionStatementSyntax declarationSyntax, SyntaxToken declaredIdentifier)
+        private T GetRemappedSymbol<T>(T originalSymbol) where T : Symbol
         {
-            return GetDeclaredLocalFunction(this.GetEnclosingBinder(GetAdjustedNodePosition(declarationSyntax)), declaredIdentifier);
+            if (!Compilation.NullableSemanticAnalysisEnabled) return originalSymbol;
+
+            EnsureNullabilityAnalysisPerformedIfNecessary();
+            if (_lazyRemappedSymbols is null) return originalSymbol;
+
+            if (_lazyRemappedSymbols.TryGetValue(originalSymbol, out Symbol remappedSymbol))
```

Comment:
Minor point: Consider merging to `if (_lazyRemappedSymbols?.TryGetValue(...) == true)`.

Reply:
I specifically didn't do this because definite assignment cannot figure out that `remappedSymbol` is definitely assigned by that test.

---
In reply to: [331618450](https://github.com/dotnet/roslyn/pull/38769#discussion_r331618450) [](ancestors = 331618450)